sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DisposableMemberInNonDisposableClass.cs

```diff
@@ -28,102 +27,86 @@
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     [Rule(DiagnosticId)]
     public sealed class DisposableMemberInNonDisposableClass : SonarDiagnosticAnalyzer
     {
-        internal const string DiagnosticId = "S2931";
+        private const string DiagnosticId = "S2931";
         private const string MessageFormat = "Implement 'IDisposable' in this class and use the 'Dispose' method to call 'Dispose' on {0}.";
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
 
-
-        protected override void Initialize(SonarAnalysisContext context)
-        {
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSymbolAction(
                 analysisContext =>
                 {
-                    var namedType = analysisContext.Symbol as INamedTypeSymbol;
-                    if (!namedType.IsClass() || namedType.Implements(KnownType.System_IDisposable))
+                    var namedType = (INamedTypeSymbol)analysisContext.Symbol;
+                    if (ShouldExclude(namedType))
                     {
                         return;
                     }
-                    var disposableFields = namedType.GetMembers().OfType<IFieldSymbol>()
-                        .Where(fs => fs.IsNonStaticNonPublicDisposableField(analysisContext.Compilation.GetLanguageVersion())).ToHashSet();
-
-                    var disposableFieldsWithInitializer = disposableFields
-                        .Where(f => IsOwnerSinceDeclaration(f));
 
-                    var otherInitializationsOfFields = namedType.GetMembers().OfType<IMethodSymbol>().
-                        SelectMany(m => GetAssignementsToFieldsIn(m, analysisContext.Compilation)).
-                        Where(f => disposableFields.Contains(f));
-
-                    var message = string.Join(", ",
-                        disposableFieldsWithInitializer.
-                        Union(otherInitializationsOfFields).
-                        Distinct().
-                        Select(symbol => $"'{symbol.Name}'").
-                        OrderBy(s => s));
+                    var message = GetMessage(namedType, analysisContext);
+                    if (string.IsNullOrEmpty(message))
+                    {
+                        return;
+                    }
 
-                    if (!string.IsNullOrEmpty(message))
+                    var typeDeclarations = new CSharpRemovableDeclarationCollector(namedType, analysisContext.Compilation).TypeDeclarations;
+                    foreach (var classDeclaration in typeDeclarations)
```

Comment:
for my education: I am looking at this loop and cannot understand why there must be more type declarations. Do you know why? If yes, is there an existing test case for it? Where we raise in multiple locations?

(I know it's not related to this change but I am trying to find out if you can simplify this in this PR)

Reply:
> I am looking at this loop and cannot understand why there must be more type declarations. Do you know why?

Partial classes can have multiple type declarations since they can be defined more than once.

> If yes, is there an existing test case for it? Where we raise in multiple locations?

We don't have tests for this case. I think this should be handled in a separate PR.